### DISCUSSION TITLE
Retain perceptual intent colors in relative intent output

### DISCUSSION THREAD

**Discussion by SoupyGit** [ImageMagick/ImageMagick#5963] [Digital Image Processing]
I export an image with the following profile from darktable:
sRGB_v4_ICC_preference.icc  
available from: https://www.color.org/srgbprofiles.xalter#v4pref

This is a v4 profile with srgb gamut, and perceptual rendering intent. The results are often very nice for images with large out of gamut areas. However, not all web browsers or web sites can accurately display either v4 profiles, and/or perceptual rendering intents. The only safe profile for consistent web display is v2, srgb, relative. So I try to convert my darktable export to that using the following:

magick /path/to/input.tif \
	-intent Relative -black-point-compensation -profile /path/to/sRGB-elle-V2-srgbtrc.icc \
	-quality 92 \
	/path/to/output.jpg

Which embeds the correct output profile, but displays the colors differently to the input.

Is there a way to convert to v2, srgb, relative, whilst retaining the colors of v4, srgb, perceptual? 

In other words, is there a way to make the colors given by v4 perceptual intent part of the image, and not just the intent? To actually change the rgb values of the image and not just rely on the intent to shift things?

Alternatively, does IM have its own gamut mapping command?

Linux Mint 21
Version: ImageMagick 7.1.0-43 Q16-HDRI x86_64 8718d6292:20220709

**Comment by snibgo** [comment]
IM doesn't have its own gamut mapping command [*]. This operation is done by `-profile`, which uses LCMS, and uses one of the four intents. However, not all ICC colour profiles support all intents.

Your input.tif has an embedded sRGB profile. Assuming your pixels values are all within gamut, then converting to another sRGB profile without `-black-point-compensation` should not change pixel values or colours, aside from very minor rounding errors that would not be visible.

If you remove `-black-point-compensation`, do you still have a noticeable change?

> In other words, is there a way to make the colors given by v4 perceptual intent part of the image, and not just the intent? To actually change the rgb values of the image and not just rely on the intent to shift things?

Yes, this is what should happen. `-profile` will change pixel values so they are all within gamut. The algorithm used for the change depends on the `-intent` setting.

* EDIT: We can use various operations that put pixel values that are out-of-gamut inside that gamut. These include `-clamp` (simple clipping) and `-auto-level' (shrink values so the minimum is zero and the maximum is 100%), and more complex filmic-like curve operations.

**Comment by SoupyGit** [comment]
Thanks for your reply. Turning bpc on/off makes a difference, but doesn't give an exact match to the perceptual output.

Here is the darktable export using their srgb (websafe) profile, which I believe is srgb, v2, relative intent:

![guys-i-discovered-led-5D3_9253-srgb-v2-relative](https://user-images.githubusercontent.com/69561443/212246434-29a5f87d-ee51-4bc7-bd44-a3d51520d7ca.png)

Here is the darktable export using the profile linked in my original post, srgb, v4, perceptual intent:

![guys-i-discovered-led-5D3_9253-srgb-v4-perceptual](https://user-images.githubusercontent.com/69561443/212246605-81952c84-521b-4f23-84e7-e0ee2bd2831b.png)

Here is the IM output after running the script in original post (using png instead of tif and jpg):

![guys-i-discovered-led-5D3_9253-srgb-v4-perceptual-to-srgb-v2-relative-bpc](https://user-images.githubusercontent.com/69561443/212246818-039d90b8-f3c7-4c58-b4f1-099765307fcf.png)

And here is the IM output with bpc off:

![guys-i-discovered-led-5D3_9253-srgb-v4-perceptual-to-srgb-v2-relative](https://user-images.githubusercontent.com/69561443/212246891-cbf6b9d9-2ca2-4f4e-a25c-51e74cf3d18d.png)

The result after running the script is visually somewhere in between the first two darktable exports.
The one that looks most similar to what I see in darktable before export is v4, perceptual.

Link to original raw can be found here: https://discuss.pixls.us/t/guys-i-just-discovered-leds/28404
xmp for this edit (remove the .txt): [guys-i-discovered-led-5D3_9253.CR2.xmp.txt](https://github.com/ImageMagick/ImageMagick/files/10408592/guys-i-discovered-led-5D3_9253.CR2.xmp.txt)

**Comment by snibgo** [comment]
Comparing the final two images, IM output with or without BPC, I can see a difference in lightness but not in colours. This seems reasonable.

The four images you show are PNG format, which cannot record out-of-gamut colours. If I convert the first to a different sRGB profile, the change is virtually zero:
```
%IMG7%magick ^
  soupy1.png ^
  ( +clone ^
    -profile %ICCPROF%\sRGB-elle-V4-srgbtrc.icc ^
    +write x.png ^
  ) ^
  -metric RMSE ^
  -format "%[distortion]\n" ^
  -compare ^
  info:

2.92986e-05
```
On a scale of 0.0 (no change) to 1.0 (maximum change), 0.0000292 is virtually zero. Visually, I can't see any difference between soupy1.png and x.png.

You made these images from an export from darktable. Is that a tiff file? Does it record out-of-gamut colours? If it does, then converting that with different sRGB profiles may give different colours, because they may have different algorithms for pushing colours inside the gamut.

If it is a tiff, you can zip it and post that to this forum.

We can use dcraw to convert the raw file 5D3_9253.CR2 to XYZ colorspace, and use IM to convert that to xyY, and plot the xy values on a CIE chromaticity diagram:

![sxs_xyy](https://user-images.githubusercontent.com/33812368/212339940-d80831c7-b360-46e5-b165-ffdca1c0eb77.png)

The small circles are the sRGB primaries and whitepoint. Many colours are outside sRGB, and even outside the horseshoe so they are apparently not colours at all. Someone on pixls.us once explained this to me: the camera matrix is optimized for minimum error with "normal photographic" colours, so highly saturated colours have high errors.

This make me wonder if a different camera matrix would do a better job, or perhaps we should use a non-matrix approach.

**Comment by SoupyGit** [comment]
> If I convert the first to a different sRGB profile, the change is virtually zero:

All you are doing there is converting a srgb v2 relative profile, to a srgb v4 relative profile. So virtually no change is expected. But that does not address the original question, which is converting srgb v4 perceptual profile, to srgb v2 relative profile. For that, you will need to try converting the second png I posted, not the first. The result you get will look like the third image (bpc on) or the fourth (bpc off).

> You made these images from an export from darktable. Is that a tiff file?

darktable itself works on the raw. The images I posted above were exported from dt exactly as you see them - png's.
I have since tried the experiment by exporting tiff's instead, with exactly the same results.

> Does it record out-of-gamut colours?

The out of gamut colours are handled by the intent.

> If it does, then converting that with different sRGB profiles may give different colours, because they may have different algorithms for pushing colours inside the gamut.

Yes, this is what's happening. Perceptual intent pushes the colours inside the gamut differently to relative intent.

However, once those colours have been pushed in gamut, they shouldn't need to be handled by the intent a second time, because they are no longer OOG. Yet given exports 3 and 4 above differ from export 2, this shows those colours are still being handled by the intent. So the question: Using IM, is there a way to bake those now in gamut colours into the rgb values, so that when it comes to export from IM with the srgb v2 relative profile, they are not handled by the intent?

> Many colours are outside sRGB, and even outside the horseshoe so they are apparently not colours at all.

Yes indeed. If that wasn't the case, this whole discussion wouldn't have started.

> This make me wonder if a different camera matrix would do a better job, or perhaps we should use a non-matrix approach.

This is not the problem I am trying to address. I am not concerned about this particular image, or this particular camera. What I am concerned about is the best method to map OOG colours to In gamut colours for final display. Indeed, the profile I used to achieve best colours here was a srgb v4 perceptual profile. I believe the perceptual profile utilises a LUT, unlike the relative profile, which is matrix. So it is a non-matrix approach (albeit at the output stage, not the input).

The problem is that not all browsers/webpages can properly display v4 perceptual profiles. However they can properly display v2 relative profiles. So what I want is the v4 perceptual colours, in a v2 relative profile. Theoretically it should be possible, since once the perceptual intent is applied, the colours are now in gamut. So now, the problem is not that they are OOG, the problem is that they are still being handled by the intent. At this point, what I want is for those colours to be baked into the image, so they are no longer handled by the intent. Then there should be no difference when converting to v2 relative.

When I used photoshop they had a thing called a stamp visible layer. It worked like this. Lets say I had a white background, and three different layers of text above. If I saved that as a .psd and sent it to someone else, they would only see the same result as me if they had the same fonts installed on their computer. However, if after the three different layers of text I made a stamp visible layer on top, saved that as .psd, and sent it, then they would see exactly the same as me (however, the text would not be editable on the stamp visible layer, but would on the three layers underneath). The stamp visible layer was a rasterized version of all that was visible. In other words, what you see is what you get. A bit like exporting, then loading the exported image on top as a layer.

I was hoping IM might be able to do something similar here. Make a stamp visible layer of the srgb v4 perceptual png. That way, all the colours would be baked into the rgb values, and no longer handled by the intent. In script it would look like this:

```
magick /path/to/v4-perceptual-input.png \
- STAMP VISIBLE \
-intent Relative -black-point-compensation -profile /path/to/sRGB-elle-V2-srgbtrc.icc \
/path/to/v2-relative-output.png
```

**Comment by snibgo** [comment]
> What I am concerned about is the best method to map OOG colours to In gamut colours for final display.

But the PNG files have no OOG colours, because PNG cannot record values outside 0 to 100%.

>  So now, the problem is not that they are OOG, the problem is that they are still being handled by the intent. At this point, what I want is for those colours to be baked into the image, so they are no longer handled by the intent.

As far as I know, the intent influences the calculation that is applied to the colours when making the output image. So, after doing "`-intent XXX -profile YYY.icc`", the colours are baked in to the image. The intent that was used is stored in the metadata, but I suppose it is relevant only if the image is eventually converted to yet another profile. However, I don't fully understand colorspaces.

**Comment by SoupyGit** [comment]
Thanks. A bit of a mystery. The closest I have come now is this:

```
magick /path/to/input.png \
-intent perceptual -profile /path/to/sRGB-elle-V2-srgbtrc.icc \
/path/to/output.png
```

Which retains mostly the same lightness but still has a shift from red towards orange.

![guys-i-discovered-led-5D3_9253-srgb-v4-perceptual-to-v2-latest-no-bpc](https://user-images.githubusercontent.com/69561443/212459921-eea2c4f9-6256-47a6-90a5-b2cb2653e94d.png)
